src/components/TableDialog.vue

```diff
@@ -157,6 +187,88 @@ export default Vue.extend({
         this.fileName = this.fileName || getFileName(file);
         this.fileUploadError = null;
       }
+
+      interface TypeScore {
+        strings: Set<string>;
+        number: number;
+        date: number;
+        total: number;
+      }
+      const columnTypes = new Map<string, TypeScore>();
+
+      Papa.parse(file, {
+        header: true,
+        skipEmptyLines: true,
+        step(row: ParseResult<{}>) {
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const data = row.data as { [key: string]: any };
+
+          Object.keys(data).forEach((key: string) => {
+            if (!columnTypes.has(key)) {
+              columnTypes.set(key, {
+                strings: new Set<string>(),
+                number: 0,
+                date: 0,
+                total: 0,
+              });
+            }
+
+            const entry = columnTypes.get(key);
+            if (entry === undefined) {
+              throw new Error('impossible');
+            }
+
+            // Pass the value through a gauntlet of heuristics to see what types
+            // it can convert to.
+            const value = data[key];
+            entry.total += 1;
+
+            // Test for categoriness by counting up all the unique strings in
+            // the column.
+            entry.strings.add(value);
+
+            // See if the value can be converted to a number. Empty strings
+            // convert to 0, so those are excluded specifically.
+            if (value !== '' && !Number.isNaN(Number(value))) {
```

Comment:
It really doesn't matter but the only way for a string to be falsy is if it's empty, so you could do this
```suggestion
            if (value && !Number.isNaN(Number(value))) {
```

Reply:
That's not true (I think Roni mentioned this on slack). The string '0' is falsy

Reply:
I don't think that's true. If you evaluate the expression `!!'0'` in your browser console, the result is `true`. https://developer.mozilla.org/en-US/docs/Glossary/Falsy

Reply:
I stand corrected. I thought that `'0' == false` (which evaluates to true) implied falsy, but that's just js doing its thing with the double equals. Your change seems appropriate, @AlmightyYakob

Reply:
Nice catch. I think the exact confusion in this thread is the reason I like to be more explicit with this sort of test. I think it leaves the test more comprehensible by future readers as well.

However, you've made me realize there's a bug in the code: I should really be working with `value.trim()` instead of `value` to avoid having to worry about whitespace-only strings (which convert to `0` but are truthy).

I will add a commit to trim the value we're working with here.

Reply:
See 21a337222bfea804dc0734e5b7bba5a6132d7643.